nexus/src/lib.rs

```diff
@@ -417,6 +418,46 @@ pub async fn run_server(config: &NexusConfig) -> Result<(), String> {
     }
     let internal_server = InternalServer::start(config, &log).await?;
     let server = Server::start(internal_server).await?;
-    server.register_as_producer().await;
     server.wait_for_finish().await
 }
+
+/// Create a new metric producer server.
+fn start_producer_server(
+    log: &Logger,
+    registry: &ProducerRegistry,
+    nexus_addr: SocketAddr,
+) -> Result<ProducerServer, String> {
+    // The producer server should listen on any available port, using the
+    // same IP as the main Dropshot server.
+    let address = SocketAddr::new(nexus_addr.ip(), 0);
```

Comment:
I don't object to this, but just making sure I understand - this is a fairly significant change to producers, right? Now every time the producer restarts it will listen on a different port, so _every_ producer is now effectively ephemeral and dependent on the garbage collection task to clean up past instances of itself?

Should this PR also remove the `disabled` bit from the gc task? Or do the producers outside of `omicron` need to be updated first to reregister themselves?

Reply:
> I don't object to this, but just making sure I understand - this is a fairly significant change to producers, right? Now every time the producer restarts it will listen on a different port

This part is correct. Some producers, like Nexus, previously attached their metric endpoints to their "main" Dropshot server. One of the big changes in this PR is that that's not supported anymore. You _have_ to run a new Dropshot server. The point of that is that it provides a central place to run the periodic registration task, inside the `oximeter_producer::Server`.

> so every producer is now effectively ephemeral and dependent on the garbage collection task to clean up past instances of itself?

I don't think this part follows. Suppose a producer restarts, so that it is now running the Dropshot server on a new port. The next time it registers with Nexus, that will _update_ the database record for that producer. There is no past instance of itself, there's only one record. (This is assuming the producer is restarted with the same UUID! That's true for Nexus, and if it wasn't true before, then this was always an issue. This PR doesn't make it worse, unless I'm missing something.)

> Should this PR also remove the disabled bit from the gc task? Or do the producers outside of omicron need to be updated first to reregister themselves?

I think we need to update all the out-of-tree consumers to use this new producer server first. Once that's rolled out, we can turn on the GC task.

Reply:
> I don't think this part follows. Suppose a producer restarts, so that it is now running the Dropshot server on a new port. The next time it registers with Nexus, that will _update_ the database record for that producer. There is no past instance of itself, there's only one record. (This is assuming the producer is restarted with the same UUID! That's true for Nexus, and if it wasn't true before, then this was always an issue. This PR doesn't make it worse, unless I'm missing something.)

Ah hah, yes, I was forgetting/missing that it restarts with the same ID. Thanks.

> I think we need to update all the out-of-tree consumers to use this new producer server first. Once that's rolled out, we can turn on the GC task.

:+1: